Unsafe macros and #![forbid(unsafe_code)]

There are macros like `unsafe_unpinned!`, which are documented as unsafe and creates a safe function, but with an unsafe block inside.

Does it mean that a crate under `#![forbid(unsafe_code)]` rule may circumvent the safety by linking to pin_utils? Shall those macros somehow require `unsafe{}` block or define an unsafe function?

Or should entire "pin_utils" be an "unsafe crate", not directly linkable from `#![forbid(unsafe_code)]` crates?

What if just make the macro insert `#[allow(unsafe_code)]`, which may fail with `error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)`? Will it fail in in forbid-unsafe-code mode?

On the other hand, it may contradict with user intentions who use `deny(unsafe_code)` instead of `forbid(unsafe_code)` to make almost-no-unsafe environment except of just one block, specifically marked with `allow(unsafe_code)` permit.

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,14 +21,10 @@ matrix:
       os: linux
 
     - name: cargo clippy
-      rust: nightly
+      rust: stable
       script:
-        - if rustup component add clippy-preview;
-          then
-            cargo clippy -- -Dwarnings;
-          else
-            echo 'Skipping clippy';
-          fi
+        - rustup component add clippy
+        - cargo clippy -- -Dwarnings
 
     - name: cargo doc
       rust: nightly

--- a/src/projection.rs
+++ b/src/projection.rs
@@ -35,6 +35,7 @@
 #[macro_export]
 macro_rules! unsafe_pinned {
     ($f:tt: $t:ty) => (
+        #[allow(unsafe_code)]
         fn $f<'__a>(
             self: $crate::core_reexport::pin::Pin<&'__a mut Self>
         ) -> $crate::core_reexport::pin::Pin<&'__a mut $t> {
@@ -77,6 +78,7 @@ macro_rules! unsafe_pinned {
 #[macro_export]
 macro_rules! unsafe_unpinned {
     ($f:tt: $t:ty) => (
+        #[allow(unsafe_code)]
         fn $f<'__a>(
             self: $crate::core_reexport::pin::Pin<&'__a mut Self>
         ) -> &'__a mut $t {

--- a/tests/stack_pin.rs
+++ b/tests/stack_pin.rs
@@ -1,3 +1,5 @@
+#![forbid(unsafe_code)] // pin_mut! is completely safe.
+
 use pin_utils::pin_mut;
 use core::pin::Pin;
 

